core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -160,7 +171,18 @@ public AllocationService(Settings settings, AllocationDeciders allocationDecider
         if (!reroute(allocation)) {
             return new RoutingAllocation.Result(false, clusterState.routingTable());
         }
-        return new RoutingAllocation.Result(true, new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        RoutingAllocation.Result result = new RoutingAllocation.Result(true,
+                new RoutingTable.Builder().updateNodes(routingNodes).build().validateRaiseException(clusterState.metaData()));
+        logClusterHealthStateChange(clusterState, ClusterState.builder(clusterState).routingResult(result).build(), "rerouting");
```

Comment:
it would be great if we can get more info about why this reroute happened. Maybe add a String reason parameter to the reroute method? This is called in many places for different reasons. (node left/join/shard store fetched etc..)